// @wireapp/store-engine/dist/commonjs/engine/MemoryEngine.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("./error/");
class MemoryEngine {
    constructor() {
        this.storeName = '';
        this.stores = {};
        this.autoIncrementedPrimaryKey = 1;
    }
    create(tableName, primaryKey, entity) {
        if (entity) {
            this.prepareTable(tableName);
            if (primaryKey === undefined) {
                primaryKey = this.autoIncrementedPrimaryKey;
                this.autoIncrementedPrimaryKey += 1;
            }
            const record = this.stores[this.storeName][tableName][primaryKey];
            if (record) {
                const message = `Record "${primaryKey}" already exists in "${tableName}". You need to delete the record first if you want to overwrite it.`;
                const error = new error_1.RecordAlreadyExistsError(message);
                return Promise.reject(error);
            }
            this.stores[this.storeName][tableName][primaryKey] = entity;
            return Promise.resolve(primaryKey);
        }
        const message = `Record "${primaryKey}" cannot be saved in "${tableName}" because it's "undefined" or "null".`;
        return Promise.reject(new error_1.RecordTypeError(message));
    }
    clearTables() {
        return __awaiter(this, void 0, void 0, function* () {
            const tableNames = Object.keys(this.stores);
            yield Promise.all(tableNames.map(tableName => this.deleteAll(tableName)));
        });
    }
    delete(tableName, primaryKey) {
        return __awaiter(this, void 0, void 0, function* () {
            this.prepareTable(tableName);
            delete this.stores[this.storeName][tableName][primaryKey];
            return primaryKey;
        });
    }
    deleteAll(tableName) {
        return __awaiter(this, void 0, void 0, function* () {
            delete this.stores[this.storeName][tableName];
            return true;
        });
    }
    init(storeName) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.assignDb(storeName, {});
        });
    }
    initWithObject(storeName, object) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.assignDb(storeName, object);
        });
    }
    assignDb(storeName, object) {
        this.storeName = storeName;
        this.stores[this.storeName] = this.stores[this.storeName] || object;
        return this.stores;
    }
    isSupported() {
        return __awaiter(this, void 0, void 0, function* () {
            // Always available
        });
    }
    purge() {
        return __awaiter(this, void 0, void 0, function* () {
            delete this.stores[this.storeName];
        });
    }
    read(tableName, primaryKey) {
        this.prepareTable(tableName);
        if (this.stores[this.storeName][tableName].hasOwnProperty(primaryKey)) {
            return Promise.resolve(this.stores[this.storeName][tableName][primaryKey]);
        }
        const message = `Record "${primaryKey}" in "${tableName}" could not be found.`;
        return Promise.reject(new error_1.RecordNotFoundError(message));
    }
    readAll(tableName) {
        this.prepareTable(tableName);
        const promises = [];
        for (const primaryKey of Object.keys(this.stores[this.storeName][tableName])) {
            promises.push(this.read(tableName, primaryKey));
        }
        return Promise.all(promises);
    }
    readAllPrimaryKeys(tableName) {
        this.prepareTable(tableName);
        return Promise.resolve(Object.keys(this.stores[this.storeName][tableName]));
    }
    update(tableName, primaryKey, changes) {
        return __awaiter(this, void 0, void 0, function* () {
            this.prepareTable(tableName);
            const entity = yield this.read(tableName, primaryKey);
            const updatedEntity = Object.assign(Object.assign({}, entity), changes);
            this.stores[this.storeName][tableName][primaryKey] = updatedEntity;
            return primaryKey;
        });
    }
    updateOrCreate(tableName, primaryKey, changes) {
        return __awaiter(this, void 0, void 0, function* () {
            this.prepareTable(tableName);
            try {
                yield this.update(tableName, primaryKey, changes);
                return primaryKey;
            }
            catch (error) {
                if (error instanceof error_1.RecordNotFoundError) {
                    const newPrimaryKey = yield this.create(tableName, primaryKey, changes);
                    return newPrimaryKey;
                }
                throw error;
            }
        });
    }
    prepareTable(tableName) {
        if (!this.stores[this.storeName][tableName]) {
            this.stores[this.storeName][tableName] = {};
        }
    }
}
exports.MemoryEngine = MemoryEngine;
