// @wireapp/proteus/src/errors/DecodeError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecodeError = void 0;
const ProteusError_1 = require("./ProteusError");
class DecodeError extends ProteusError_1.ProteusError {
    constructor(message = 'Unknown decoding error', code = 3) {
        super(message, code);
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.DecodeError = DecodeError;
DecodeError.CODE = {
    CASE_300: 300,
    CASE_301: 301,
    CASE_302: 302,
    CASE_303: 303,
};
(function (DecodeError) {
    class InvalidType extends DecodeError {
        constructor(message = 'Invalid type', code) {
            super(message, code);
            Object.setPrototypeOf(this, new.target.prototype);
        }
    }
    DecodeError.InvalidType = InvalidType;
    class InvalidArrayLen extends DecodeError {
        constructor(message = 'Invalid array length', code) {
            super(message, code);
            Object.setPrototypeOf(this, new.target.prototype);
        }
    }
    DecodeError.InvalidArrayLen = InvalidArrayLen;
    class LocalIdentityChanged extends DecodeError {
        constructor(message = 'Local identity changed', code) {
            super(message, code);
            Object.setPrototypeOf(this, new.target.prototype);
        }
    }
    DecodeError.LocalIdentityChanged = LocalIdentityChanged;
})(DecodeError = exports.DecodeError || (exports.DecodeError = {}));
