// @wireapp/proteus/src/errors/InputError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputError = void 0;
const ProteusError_1 = require("./ProteusError");
class InputError extends ProteusError_1.ProteusError {
    constructor(message = 'Invalid input', code = 4) {
        super(message, code);
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.InputError = InputError;
InputError.CODE = {
    CASE_400: 400,
    CASE_401: 401,
    CASE_402: 402,
    CASE_403: 403,
    CASE_404: 404,
    CASE_405: 405,
    CASE_406: 406,
    CASE_407: 407,
    CASE_408: 408,
    CASE_409: 409,
};
(function (InputError) {
    class RangeError extends InputError {
        constructor(message = 'Invalid array length', code) {
            super(message, code);
            Object.setPrototypeOf(this, new.target.prototype);
        }
    }
    InputError.RangeError = RangeError;
    class TypeError extends InputError {
        constructor(message = 'Invalid type', code) {
            super(message, code);
            Object.setPrototypeOf(this, new.target.prototype);
        }
    }
    InputError.TypeError = TypeError;
    class ConversionError extends InputError {
        constructor(message = 'Conversion error', code) {
            super(message, code);
            Object.setPrototypeOf(this, new.target.prototype);
        }
    }
    InputError.ConversionError = ConversionError;
})(InputError = exports.InputError || (exports.InputError = {}));
