// bazinga64/dist/test/TestValues.js
"use strict";
/*
 * Wire
 * Copyright (C) 2019 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.numberString = exports.numberEncoded = exports.numberDecoded = exports.helloEncodedString = exports.helloEncodedArray = exports.helloDecodedString = exports.helloDecodedArray = void 0;
exports.helloDecodedArray = [72, 101, 108, 108, 111];
exports.helloDecodedString = 'Hello';
exports.helloEncodedArray = [83, 71, 86, 115, 98, 71, 56, 61];
exports.helloEncodedString = 'SGVsbG8=';
exports.numberDecoded = 1337;
exports.numberEncoded = 'MTMzNw==';
exports.numberString = '1337';
