// @wireapp/priority-queue/src/main/Item.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Item = void 0;
const Priority_1 = require("./Priority");
class Item {
    constructor() {
        /** original business logic */
        this.fn = () => { };
        this.priority = Priority_1.Priority.MEDIUM;
        /** wrapped `reject` of `fn` */
        this.reject = () => { };
        /** wrapped `resolve` of `fn` */
        this.resolve = () => { };
        /** number of remaining retries for rejecting Promises */
        this.retry = Infinity;
        /** time when the item has been added to the queue */
        this.timestamp = 0;
    }
    toString() {
        return `
    label=${this.label},
    priority=${this.priority},
    timestamp=${this.timestamp},
    retry=${this.retry},
    fn=${this.fn.toString().replace(/(\r\n|\n|\r|\s+)/gm, ' ')},`;
    }
}
exports.Item = Item;
