// @wireapp/api-client/dist/team/member/Role.js
"use strict";
/*
 * Wire
 * Copyright (C) 2019 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
const Permissions_1 = require("./Permissions");
var Role;
(function (Role) {
    Role["ADMIN"] = "admin";
    Role["EXTERNAL"] = "partner";
    Role["MEMBER"] = "member";
    Role["OWNER"] = "owner";
})(Role = exports.Role || (exports.Role = {}));
exports.roleToPermissions = (role) => {
    switch (role.toLowerCase()) {
        case Role.OWNER: {
            return Permissions_1.combinePermissions([
                exports.roleToPermissions(Role.ADMIN),
                Permissions_1.Permissions.DELETE_TEAM,
                Permissions_1.Permissions.GET_BILLING,
                Permissions_1.Permissions.SET_BILLING,
            ]);
        }
        case Role.ADMIN: {
            return Permissions_1.combinePermissions([
                exports.roleToPermissions(Role.MEMBER),
                Permissions_1.Permissions.ADD_TEAM_MEMBER,
                Permissions_1.Permissions.REMOVE_TEAM_MEMBER,
                Permissions_1.Permissions.SET_MEMBER_PERMISSIONS,
                Permissions_1.Permissions.SET_TEAM_DATA,
            ]);
        }
        case Role.MEMBER: {
            return Permissions_1.combinePermissions([
                exports.roleToPermissions(Role.EXTERNAL),
                Permissions_1.Permissions.ADD_CONVERSATION_MEMBER,
                Permissions_1.Permissions.DELETE_CONVERSATION,
                Permissions_1.Permissions.GET_MEMBER_PERMISSIONS,
                Permissions_1.Permissions.REMOVE_CONVERSATION_MEMBER,
            ]);
        }
        case Role.EXTERNAL: {
            return Permissions_1.combinePermissions([Permissions_1.Permissions.CREATE_CONVERSATION, Permissions_1.Permissions.GET_TEAM_CONVERSATIONS]);
        }
        default: {
            return 0;
        }
    }
};
exports.permissionsToRole = (permissions) => {
    if (Permissions_1.hasPermissions(permissions.self, exports.roleToPermissions(Role.OWNER))) {
        return Role.OWNER;
    }
    if (Permissions_1.hasPermissions(permissions.self, exports.roleToPermissions(Role.ADMIN))) {
        return Role.ADMIN;
    }
    if (Permissions_1.hasPermissions(permissions.self, exports.roleToPermissions(Role.MEMBER))) {
        return Role.MEMBER;
    }
    if (Permissions_1.hasPermissions(permissions.self, exports.roleToPermissions(Role.EXTERNAL))) {
        return Role.EXTERNAL;
    }
    return undefined;
};
exports.isPartner = (permissions) => {
    return !!(permissions && exports.permissionsToRole(permissions) === Role.EXTERNAL);
};
exports.isMember = (permissions) => {
    return !!(permissions && exports.permissionsToRole(permissions) === Role.MEMBER);
};
exports.isAdmin = (permissions) => {
    return !!(permissions && exports.permissionsToRole(permissions) === Role.ADMIN);
};
exports.isOwner = (permissions) => {
    return !!(permissions && exports.permissionsToRole(permissions) === Role.OWNER);
};
exports.isAtLeastPartner = (permissions) => {
    return exports.isPartner(permissions) || exports.isMember(permissions) || exports.isAdmin(permissions) || exports.isOwner(permissions);
};
exports.isAtLeastMember = (permissions) => {
    return exports.isMember(permissions) || exports.isAdmin(permissions) || exports.isOwner(permissions);
};
exports.isAtLeastAdmin = (permissions) => {
    return exports.isAdmin(permissions) || exports.isOwner(permissions);
};
exports.isAtLeastEqual = (permissions, otherPermissions) => {
    return ((exports.isOwner(permissions) && exports.isOwner(otherPermissions)) ||
        (exports.isAtLeastAdmin(permissions) && exports.isAdmin(otherPermissions)) ||
        (exports.isAtLeastMember(permissions) && exports.isMember(otherPermissions)) ||
        (exports.isAtLeastPartner(permissions) && exports.isPartner(otherPermissions)));
};
