// @wireapp/core/src/main/broadcast/BroadcastService.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BroadcastService = void 0;
const protocol_messaging_1 = require("@wireapp/protocol-messaging");
const MessageService_1 = require("../conversation/message/MessageService");
class BroadcastService {
    constructor(apiClient, cryptographyService) {
        this.apiClient = apiClient;
        this.cryptographyService = cryptographyService;
        this.messageService = new MessageService_1.MessageService(this.apiClient, this.cryptographyService);
    }
    async getPreKeyBundlesFromTeam(teamId, skipOwnClients = false) {
        const { members: teamMembers } = await this.apiClient.teams.member.api.getAllMembers(teamId);
        let members = teamMembers.map(member => ({ id: member.user }));
        if (skipOwnClients) {
            const selfUser = await this.apiClient.self.api.getSelf();
            members = members.filter(member => member.id !== selfUser.id);
        }
        const preKeys = await Promise.all(members.map(member => this.apiClient.user.api.getUserPreKeys(member.id)));
        return preKeys.reduce((bundleMap, bundle) => {
            bundleMap[bundle.user] = {};
            for (const client of bundle.clients) {
                bundleMap[bundle.user][client.client] = client.prekey;
            }
            return bundleMap;
        }, {});
    }
    async broadcastGenericMessage(genericMessage, preKeyBundles, sendAsProtobuf) {
        const plainTextArray = protocol_messaging_1.GenericMessage.encode(genericMessage).finish();
        const recipients = await this.cryptographyService.encrypt(plainTextArray, preKeyBundles);
        return sendAsProtobuf
            ? this.messageService.sendOTRProtobufMessage(this.apiClient.validatedClientId, recipients, null, plainTextArray)
            : this.messageService.sendOTRMessage(this.apiClient.validatedClientId, recipients, null, plainTextArray);
    }
}
exports.BroadcastService = BroadcastService;
