// @wireapp/api-client/dist/team/index.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var invitation_1 = require("./invitation/");
exports.TeamInvitationAPI = invitation_1.TeamInvitationAPI;
var legalhold_1 = require("./legalhold/");
exports.LegalHoldAPI = legalhold_1.LegalHoldAPI;
var member_1 = require("./member/");
exports.MemberAPI = member_1.MemberAPI;
exports.Permissions = member_1.Permissions;
exports.Role = member_1.Role;
var team_1 = require("./team/");
exports.TeamAPI = team_1.TeamAPI;
var payment_1 = require("./payment/");
exports.PaymentAPI = payment_1.PaymentAPI;
var service_1 = require("./service/");
exports.ServiceAPI = service_1.ServiceAPI;
var TeamError_1 = require("./TeamError");
exports.TeamError = TeamError_1.TeamError;
exports.InviteEmailInUseError = TeamError_1.InviteEmailInUseError;
exports.InvalidInvitationCodeError = TeamError_1.InvalidInvitationCodeError;
exports.ServiceNotFoundError = TeamError_1.ServiceNotFoundError;
__export(require("./identityprovider/"));
__export(require("./feature/"));
__export(require("./conversation"));
