// @wireapp/api-client/dist/team/service/ServiceAPI.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class ServiceAPI {
    constructor(client) {
        this.client = client;
    }
    getServices(limit = 100, start, tags) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = {
                method: 'get',
                params: {
                    size: limit,
                    start,
                    tags,
                },
                url: `/${ServiceAPI.URL.SERVICES}`,
            };
            const response = yield this.client.sendJSON(config);
            return response.data;
        });
    }
    getTeamServices(teamId, limit = 100) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = {
                method: 'get',
                params: {
                    prefix: undefined,
                    size: limit,
                    start: undefined,
                },
                url: `${ServiceAPI.URL.TEAMS}/${teamId}/${ServiceAPI.URL.SERVICES}/${ServiceAPI.URL.WHITELISTED}`,
            };
            const response = yield this.client.sendJSON(config);
            return response.data;
        });
    }
    getProvider(providerId) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = {
                method: 'get',
                url: `${ServiceAPI.URL.PROVIDERS}/${providerId}`,
            };
            const response = yield this.client.sendJSON(config);
            return response.data;
        });
    }
    getProviderServices(providerId) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = {
                method: 'get',
                url: `${ServiceAPI.URL.PROVIDERS}/${providerId}/${ServiceAPI.URL.SERVICES}`,
            };
            const response = yield this.client.sendJSON(config);
            return response.data;
        });
    }
    getService(providerId, serviceId) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = {
                method: 'get',
                url: `${ServiceAPI.URL.PROVIDERS}/${providerId}/${ServiceAPI.URL.SERVICES}/${serviceId}`,
            };
            const response = yield this.client.sendJSON(config);
            return response.data;
        });
    }
    postServiceWhitelist(teamId, whitelistData) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = {
                data: whitelistData,
                method: 'post',
                url: `${ServiceAPI.URL.TEAMS}/${teamId}/${ServiceAPI.URL.SERVICES}/${ServiceAPI.URL.WHITELIST}`,
            };
            const response = yield this.client.sendJSON(config);
            return response.data;
        });
    }
}
exports.ServiceAPI = ServiceAPI;
ServiceAPI.URL = {
    PROVIDERS: '/providers',
    SERVICES: 'services',
    TEAMS: '/teams',
    WHITELIST: 'whitelist',
    WHITELISTED: 'whitelisted',
};
