// @wireapp/store-engine/dist/commonjs/engine/error/index.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var LowDiskSpaceError_1 = require("./LowDiskSpaceError");
exports.LowDiskSpaceError = LowDiskSpaceError_1.LowDiskSpaceError;
var PathValidationError_1 = require("./PathValidationError");
exports.PathValidationError = PathValidationError_1.PathValidationError;
var RecordAlreadyExistsError_1 = require("./RecordAlreadyExistsError");
exports.RecordAlreadyExistsError = RecordAlreadyExistsError_1.RecordAlreadyExistsError;
var RecordNotFoundError_1 = require("./RecordNotFoundError");
exports.RecordNotFoundError = RecordNotFoundError_1.RecordNotFoundError;
var RecordTypeError_1 = require("./RecordTypeError");
exports.RecordTypeError = RecordTypeError_1.RecordTypeError;
var UnsupportedError_1 = require("./UnsupportedError");
exports.UnsupportedError = UnsupportedError_1.UnsupportedError;
