// @wireapp/api-client/dist/self/SelfAPI.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfAPI = void 0;
let SelfAPI = /** @class */ (() => {
    class SelfAPI {
        constructor(client) {
            this.client = client;
        }
        /**
         * Remove your email address.
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/removeEmail
         */
        deleteEmail() {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'delete',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.EMAIL}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Remove your phone number.
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/removePhone
         */
        deletePhone() {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'delete',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.PHONE}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Initiate account deletion.
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/deleteUser
         */
        deleteSelf(deleteData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: deleteData,
                    method: 'delete',
                    url: SelfAPI.URL.SELF,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Get your profile name.
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/selfName
         */
        getName() {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.NAME}`,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        /**
         * Get your consents.
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/consent
         */
        getConsents() {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.CONSENT}`,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        /**
         * Put your consent.
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/consent
         */
        putConsent(consent) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: consent,
                    method: 'put',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.CONSENT}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Get your profile
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/self
         */
        getSelf() {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: SelfAPI.URL.SELF,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        /**
         * Change your email address.
         * @param emailData The new email address
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/changeEmail
         */
        putEmail(emailData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: emailData,
                    method: 'put',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.EMAIL}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Change your handle.
         * @param handleData The new handle
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/changeHandle
         */
        putHandle(handleData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: handleData,
                    method: 'put',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.HANDLE}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Change your locale.
         * @param localeData The new locale
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/changeLocale
         */
        putLocale(localeData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: localeData,
                    method: 'put',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.LOCALE}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Change your password.
         * @param passwordData The new password
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/changePassword
         */
        putPassword(passwordData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: passwordData,
                    method: 'put',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.PASSWORD}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Change your phone number.
         * @param phoneData The new phone number (E.164 format)
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/changePhone
         */
        putPhone(phoneData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: phoneData,
                    method: 'put',
                    url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.PHONE}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Update your profile.
         * @param profileData The new profile data
         * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/users/updateSelf
         */
        putSelf(profileData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: profileData,
                    method: 'put',
                    url: SelfAPI.URL.SELF,
                };
                yield this.client.sendJSON(config);
            });
        }
        /**
         * Check if password is set.
         */
        headPassword() {
            const config = {
                method: 'head',
                url: `${SelfAPI.URL.SELF}/${SelfAPI.URL.PASSWORD}`,
            };
            return this.client.sendJSON(config);
        }
    }
    SelfAPI.URL = {
        CONSENT: 'consent',
        EMAIL: 'email',
        HANDLE: 'handle',
        LOCALE: 'locale',
        NAME: 'name',
        PASSWORD: 'password',
        PHONE: 'phone',
        SELF: '/self',
    };
    return SelfAPI;
})();
exports.SelfAPI = SelfAPI;
