// @wireapp/priority-queue/dist/PriorityQueue.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logdown_1 = __importDefault(require("logdown"));
const Item_1 = require("./Item");
const Priority_1 = require("./Priority");
class PriorityQueue {
    constructor(config) {
        this.logger = logdown_1.default('@wireapp/priority-queue/PriorityQueue', {
            logger: console,
            markdown: false,
        });
        this.config = {
            comparator: (a, b) => {
                if (a.priority === b.priority) {
                    return a.timestamp - b.timestamp;
                }
                return b.priority - a.priority;
            },
            maxRetries: 0,
            maxRetryDelay: Number.MAX_SAFE_INTEGER,
            retryDelay: 1000,
            retryGrowthFactor: 1.3,
        };
        this.isRunning = false;
        this.queue = [];
        this.config = Object.assign(Object.assign({}, this.config), config);
    }
    add(thunkedPromise, priority = Priority_1.Priority.MEDIUM, label) {
        return new Promise((resolve, reject) => {
            const queueObject = new Item_1.Item();
            queueObject.fn = thunkedPromise;
            queueObject.label = label;
            queueObject.priority = priority;
            queueObject.reject = reject;
            queueObject.resolve = resolve;
            queueObject.retry = 0;
            queueObject.timestamp = Date.now() + this.size;
            this.queue.push(queueObject);
            this.queue.sort(this.config.comparator);
            if (!this.isRunning) {
                this.isRunning = true;
                // eslint-disable-next-line @typescript-eslint/no-floating-promises
                this.processList();
            }
        });
    }
    delete(label) {
        this.queue = this.queue.filter(item => item.label !== label);
    }
    deleteAll() {
        this.queue = [];
    }
    get all() {
        return this.queue;
    }
    get first() {
        return this.queue[0];
    }
    get last() {
        return this.queue[this.queue.length - 1];
    }
    get size() {
        return this.queue.length;
    }
    processList() {
        return __awaiter(this, void 0, void 0, function* () {
            const queueObject = this.first;
            if (!queueObject) {
                this.isRunning = false;
                return;
            }
            try {
                queueObject.resolve(yield queueObject.fn());
                this.queue.shift();
                // eslint-disable-next-line @typescript-eslint/no-floating-promises
                this.processList();
            }
            catch (error) {
                if (queueObject.retry >= this.config.maxRetries) {
                    this.queue.shift();
                    queueObject.reject(error);
                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
                    this.processList();
                }
                else {
                    this.logger.log(`Retrying item "${queueObject}"`);
                    setTimeout(() => this.processList(), this.getGrowingDelay(queueObject.retry));
                    queueObject.retry++;
                }
            }
        });
    }
    getGrowingDelay(currentRetry) {
        const delay = currentRetry < 1 ? this.config.retryDelay : this.config.retryDelay * currentRetry * this.config.retryGrowthFactor;
        return Math.min(delay, this.config.maxRetryDelay);
    }
    toString() {
        return this.queue
            .map((item, index) => {
            return `"${index}": ${item.toString()}`;
        })
            .join('\r\n');
    }
}
exports.PriorityQueue = PriorityQueue;
