// @wireapp/api-client/dist/team/legalhold/LegalHoldAPI.js
"use strict";
/*
 * Wire
 * Copyright (C) 2019 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegalHoldAPI = void 0;
const TeamAPI_1 = require("../team/TeamAPI");
let LegalHoldAPI = /** @class */ (() => {
    class LegalHoldAPI {
        constructor(client) {
            this.client = client;
        }
        getMemberLegalHold(teamId, userId) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: `${TeamAPI_1.TeamAPI.URL.TEAMS}/${teamId}/${LegalHoldAPI.URL.LEGAL_HOLD}/${userId}`,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        deleteMemberLegalHold(teamId, userId, password) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: {
                        password,
                    },
                    method: 'delete',
                    url: `${TeamAPI_1.TeamAPI.URL.TEAMS}/${teamId}/${LegalHoldAPI.URL.LEGAL_HOLD}/${userId}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        postMemberLegalHold(teamId, userId) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'post',
                    url: `${TeamAPI_1.TeamAPI.URL.TEAMS}/${teamId}/${LegalHoldAPI.URL.LEGAL_HOLD}/${userId}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        putMemberApproveLegalHold(teamId, userId, password) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: {
                        password,
                    },
                    method: 'put',
                    url: `${TeamAPI_1.TeamAPI.URL.TEAMS}/${teamId}/${LegalHoldAPI.URL.LEGAL_HOLD}/${userId}/${LegalHoldAPI.URL.APPROVE_LEGAL_HOLD}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        getSettings(teamId) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: `${TeamAPI_1.TeamAPI.URL.TEAMS}/${teamId}/${LegalHoldAPI.URL.LEGAL_HOLD}/${LegalHoldAPI.URL.SETTINGS_LEGAL_HOLD}`,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        deleteSettings(teamId, password) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: {
                        password,
                    },
                    method: 'delete',
                    url: `${TeamAPI_1.TeamAPI.URL.TEAMS}/${teamId}/${LegalHoldAPI.URL.LEGAL_HOLD}/${LegalHoldAPI.URL.SETTINGS_LEGAL_HOLD}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        postSettings(teamId, legalHoldData) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: legalHoldData,
                    method: 'post',
                    url: `${TeamAPI_1.TeamAPI.URL.TEAMS}/${teamId}/${LegalHoldAPI.URL.LEGAL_HOLD}/${LegalHoldAPI.URL.SETTINGS_LEGAL_HOLD}`,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
    }
    LegalHoldAPI.URL = {
        APPROVE_LEGAL_HOLD: 'approve',
        LEGAL_HOLD: 'legalhold',
        SETTINGS_LEGAL_HOLD: 'settings',
    };
    return LegalHoldAPI;
})();
exports.LegalHoldAPI = LegalHoldAPI;
