// @wireapp/api-client/dist/event/ConversationEvent.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONVERSATION_EVENT = void 0;
var CONVERSATION_EVENT;
(function (CONVERSATION_EVENT) {
    CONVERSATION_EVENT["ACCESS_UPDATE"] = "conversation.access-update";
    CONVERSATION_EVENT["CODE_DELETE"] = "conversation.code-delete";
    CONVERSATION_EVENT["CODE_UPDATE"] = "conversation.code-update";
    CONVERSATION_EVENT["CONNECT_REQUEST"] = "conversation.connect-request";
    CONVERSATION_EVENT["CREATE"] = "conversation.create";
    CONVERSATION_EVENT["DELETE"] = "conversation.delete";
    CONVERSATION_EVENT["MEMBER_JOIN"] = "conversation.member-join";
    CONVERSATION_EVENT["MEMBER_LEAVE"] = "conversation.member-leave";
    CONVERSATION_EVENT["MEMBER_UPDATE"] = "conversation.member-update";
    CONVERSATION_EVENT["MESSAGE_TIMER_UPDATE"] = "conversation.message-timer-update";
    CONVERSATION_EVENT["OTR_MESSAGE_ADD"] = "conversation.otr-message-add";
    CONVERSATION_EVENT["RECEIPT_MODE_UPDATE"] = "conversation.receipt-mode-update";
    CONVERSATION_EVENT["RENAME"] = "conversation.rename";
    CONVERSATION_EVENT["TYPING"] = "conversation.typing";
})(CONVERSATION_EVENT = exports.CONVERSATION_EVENT || (exports.CONVERSATION_EVENT = {}));
