// @wireapp/api-client/dist/event/TeamEvent.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TEAM_EVENT = void 0;
var TEAM_EVENT;
(function (TEAM_EVENT) {
    TEAM_EVENT["CONVERSATION_CREATE"] = "team.conversation-create";
    TEAM_EVENT["CONVERSATION_DELETE"] = "team.conversation-delete";
    TEAM_EVENT["CREATE"] = "team.create";
    TEAM_EVENT["DELETE"] = "team.delete";
    TEAM_EVENT["MEMBER_JOIN"] = "team.member-join";
    TEAM_EVENT["MEMBER_LEAVE"] = "team.member-leave";
    TEAM_EVENT["MEMBER_UPDATE"] = "team.member-update";
    TEAM_EVENT["UPDATE"] = "team.update";
})(TEAM_EVENT = exports.TEAM_EVENT || (exports.TEAM_EVENT = {}));
