// @wireapp/api-client/dist/client/ClientAPI.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientAPI = void 0;
let ClientAPI = /** @class */ (() => {
    class ClientAPI {
        constructor(client) {
            this.client = client;
        }
        postClient(newClient) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: newClient,
                    method: 'post',
                    url: ClientAPI.URL.CLIENTS,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        putClient(clientId, updatedClient) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: updatedClient,
                    method: 'put',
                    url: `${ClientAPI.URL.CLIENTS}/${clientId}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        deleteClient(clientId, password) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    data: {
                        password,
                    },
                    method: 'delete',
                    url: `${ClientAPI.URL.CLIENTS}/${clientId}`,
                };
                yield this.client.sendJSON(config);
            });
        }
        getClient(clientId) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: `${ClientAPI.URL.CLIENTS}/${clientId}`,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        getClients() {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: ClientAPI.URL.CLIENTS,
                };
                const response = yield this.client.sendJSON(config);
                return response.data;
            });
        }
        getClientPreKeys(clientId) {
            return __awaiter(this, void 0, void 0, function* () {
                const config = {
                    method: 'get',
                    url: `${ClientAPI.URL.CLIENTS}/${clientId}/prekeys`,
                };
                const response = yield this.client.sendJSON(config, true);
                return response.data;
            });
        }
    }
    ClientAPI.URL = {
        CLIENTS: '/clients',
    };
    return ClientAPI;
})();
exports.ClientAPI = ClientAPI;
