// @wireapp/api-client/src/team/feature/FeatureList.js
"use strict";
/*
 * Wire
 * Copyright (C) 2020 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FEATURE_KEY = void 0;
var FEATURE_KEY;
(function (FEATURE_KEY) {
    FEATURE_KEY["APPLOCK"] = "appLock";
    FEATURE_KEY["CLASSIFIED_DOMAINS"] = "classifiedDomains";
    FEATURE_KEY["CONFERENCE_CALLING"] = "conferenceCalling";
    FEATURE_KEY["DIGITAL_SIGNATURES"] = "digitalSignatures";
    FEATURE_KEY["FILE_SHARING"] = "fileSharing";
    FEATURE_KEY["LEGALHOLD"] = "legalhold";
    FEATURE_KEY["SEARCH_VISIBILITY"] = "searchVisibility";
    FEATURE_KEY["SSO"] = "sso";
    FEATURE_KEY["VALIDATE_SAML_EMAILS"] = "validateSAMLemails";
    FEATURE_KEY["VIDEO_CALLING"] = "videoCalling";
    FEATURE_KEY["SELF_DELETING_MESSAGES"] = "selfDeletingMessages";
})(FEATURE_KEY = exports.FEATURE_KEY || (exports.FEATURE_KEY = {}));
