// @wireapp/core/dist/self/SelfService.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class SelfService {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    checkUsername(username) {
        return __awaiter(this, void 0, void 0, function* () {
            const [availableUsername] = yield this.checkUsernames([username]);
            return !!availableUsername;
        });
    }
    checkUsernames(usernames) {
        return this.apiClient.user.api.postHandles({
            handles: usernames,
        });
    }
    getName() {
        return __awaiter(this, void 0, void 0, function* () {
            const { name } = yield this.apiClient.self.api.getName();
            return name;
        });
    }
    getSelf() {
        return this.apiClient.self.api.getSelf();
    }
    getUsername() {
        return __awaiter(this, void 0, void 0, function* () {
            const { handle } = yield this.apiClient.self.api.getSelf();
            return handle;
        });
    }
    setName(name) {
        return this.apiClient.self.api.putSelf({ name });
    }
    setUsername(username) {
        return this.apiClient.self.api.putHandle({ handle: username });
    }
}
exports.SelfService = SelfService;
