// @wireapp/store-engine/dist/commonjs/test/updateSpec.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("../engine/error");
const TABLE_NAME = 'the-simpsons';
exports.updateSpec = {
    'fails if the record does not exist.': (engine) => __awaiter(void 0, void 0, void 0, function* () {
        const PRIMARY_KEY = 'primary-key';
        const updates = {
            age: 177,
            size: {
                height: 1080,
                width: 1920,
            },
        };
        try {
            yield engine.update(TABLE_NAME, PRIMARY_KEY, updates);
            fail();
        }
        catch (error) {
            expect(error).toEqual(jasmine.any(error_1.RecordNotFoundError));
        }
    }),
    'updates an existing database record.': (engine) => __awaiter(void 0, void 0, void 0, function* () {
        const PRIMARY_KEY = 'primary-key';
        const entity = {
            name: 'Old monitor',
        };
        const updates = {
            age: 177,
            size: {
                height: 1080,
                width: 1920,
            },
        };
        const expectedAmountOfProperties = 2;
        yield engine.create(TABLE_NAME, PRIMARY_KEY, entity);
        const primaryKey = yield engine.update(TABLE_NAME, PRIMARY_KEY, updates);
        const updatedRecord = yield engine.read(TABLE_NAME, primaryKey);
        expect(updatedRecord.name).toBe(entity.name);
        expect(updatedRecord.age).toBe(updates.age);
        expect(Object.keys(updatedRecord.size).length).toBe(expectedAmountOfProperties);
        expect(updatedRecord.size.height).toBe(updates.size.height);
        expect(updatedRecord.size.width).toBe(updates.size.width);
    }),
};
