// @wireapp/api-client/src/event/UserEvent.js
"use strict";
/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.USER_EVENT = void 0;
var USER_EVENT;
(function (USER_EVENT) {
    USER_EVENT["ACTIVATE"] = "user.activate";
    USER_EVENT["CLIENT_ADD"] = "user.client-add";
    USER_EVENT["CLIENT_REMOVE"] = "user.client-remove";
    USER_EVENT["CONNECTION"] = "user.connection";
    USER_EVENT["DELETE"] = "user.delete";
    USER_EVENT["LEGAL_HOLD_DISABLE"] = "user.legalhold-disable";
    USER_EVENT["LEGAL_HOLD_ENABLE"] = "user.legalhold-enable";
    USER_EVENT["LEGAL_HOLD_REQUEST"] = "user.legalhold-request";
    USER_EVENT["PROPERTIES_DELETE"] = "user.properties-delete";
    USER_EVENT["PROPERTIES_SET"] = "user.properties-set";
    USER_EVENT["PUSH_REMOVE"] = "user.push-remove";
    USER_EVENT["UPDATE"] = "user.update";
})(USER_EVENT = exports.USER_EVENT || (exports.USER_EVENT = {}));
